store/sqlstore/group_supplier.go

```diff
@@ -0,0 +1,864 @@
+}
+
+func (s *SqlSupplier) GroupDeleteGroupSyncable(ctx context.Context, groupID string, syncableID string, syncableType model.GroupSyncableType, hints ...store.LayeredStoreHint) *store.LayeredStoreSupplierResult {
+	result := store.NewSupplierResult()
+
+	if !model.IsValidId(groupID) {
+		result.Err = model.NewAppError("SqlGroupStore.GroupDeleteGroupSyncable", "model.group.id.app_error", nil, "group_id="+groupID, http.StatusBadRequest)
+		return result
+	}
+
+	if !model.IsValidId(syncableID) {
+		result.Err = model.NewAppError("SqlGroupStore.GroupDeleteGroupSyncable", "model.group_syncable.syncable_id.app_error", nil, "group_id="+groupID, http.StatusBadRequest)
+		return result
+	}
+
+	groupSyncable, err := s.getGroupSyncable(groupID, syncableID, syncableType)
+	if err != nil {
+		if err == sql.ErrNoRows {
+			result.Err = model.NewAppError("SqlGroupStore.GroupDeleteGroupSyncable", "store.sql_group.no_rows", nil, "Id="+groupID+", "+err.Error(), http.StatusNotFound)
+		} else {
+			result.Err = model.NewAppError("SqlGroupStore.GroupDeleteGroupSyncable", "store.select_error", nil, err.Error(), http.StatusInternalServerError)
+		}
+		return result
+	}
+
+	if groupSyncable.DeleteAt != 0 {
+		result.Err = model.NewAppError("SqlGroupStore.GroupDeleteGroupSyncable", "store.sql_group.group_syncable_already_deleted", nil, "group_id="+groupID+"syncable_id="+syncableID, http.StatusBadRequest)
+		return result
+	}
+
+	time := model.GetMillis()
+	groupSyncable.DeleteAt = time
+	groupSyncable.UpdateAt = time
+
+	var rowsAffected int64
+	switch groupSyncable.Type {
+	case model.GroupSyncableTypeTeam:
+		rowsAffected, err = s.GetMaster().Update(groupSyncableToGroupTeam(groupSyncable))
+	case model.GroupSyncableTypeChannel:
+		rowsAffected, err = s.GetMaster().Update(groupSyncableToGroupChannel(groupSyncable))
+	default:
+		model.NewAppError("SqlGroupStore.GroupDeleteGroupSyncable", "model.group.type.app_error", nil, "group_id="+groupSyncable.GroupId+", syncable_id="+groupSyncable.SyncableId+", "+err.Error(), http.StatusInternalServerError)
+		return result
+	}
+
+	if err != nil {
+		result.Err = model.NewAppError("SqlGroupStore.GroupDeleteGroupSyncable", "store.update_error", nil, err.Error(), http.StatusInternalServerError)
+		return result
+	}
+
+	if rowsAffected == 0 {
+		result.Err = model.NewAppError("SqlGroupStore.GroupDeleteGroupSyncable", "store.sql_group.no_rows_affected", nil, "", http.StatusInternalServerError)
+		return result
+	}
+
+	result.Data = groupSyncable
+
+	return result
+}
+
+// PendingAutoAddTeamMembers returns a slice of UserTeamIDPair that need newly created memberships
+// based on the groups configurations.
+//
+// Typically minGroupMembersCreateAt will be the last successful group sync time.
+func (s *SqlSupplier) PendingAutoAddTeamMembers(ctx context.Context, minGroupMembersCreateAt int64, hints ...store.LayeredStoreHint) *store.LayeredStoreSupplierResult {
+	result := store.NewSupplierResult()
+
+	query := `
+		SELECT 
```

Comment:
I'm slightly worried about the performance of this query: it's always hard to predict how the query planner will ruin the party for different installation sizes, etc. I could imagine some improvements to make it more deterministic, but I suggest we ensure this is performance tested first (e.g. triggering a mock ldap sync throughout a performance test run to ensure the query doesn't lock tables, etc.)